File: projects/rpn2tex/migrations/go-feature-by-feature-1/lexer.go
package main

import "unicode"

// Lexer tokenizes input text for RPN expressions
type Lexer struct {
	text   string
	pos    int
	line   int
	column int
}

// NewLexer creates a new lexer for the given input text
func NewLexer(text string) *Lexer {
	return &Lexer{
		text:   text,
		pos:    0,
		line:   1,
		column: 1,
	}
}

// Tokenize returns all tokens from the input
func (l *Lexer) Tokenize() ([]Token, error) {
	var tokens []Token
	for !l.atEnd() {
		token, err := l.nextToken()
		if err != nil {
			return nil, err
		}
		if token != nil {
			tokens = append(tokens, *token)
		}
	}
	return tokens, nil
}

// nextToken returns the next token from the input
func (l *Lexer) nextToken() (*Token, error) {
	// Skip whitespace
	l.skipWhitespace()

	if l.atEnd() {
		return nil, nil
	}

	ch := l.peek()
	startLine := l.line
	startColumn := l.column

	// Check for numbers
	if unicode.IsDigit(ch) {
		return l.scanNumber("", startLine, startColumn), nil
	}

	// Check for minus sign (could be negative number or subtraction operator)
	if ch == '-' {
		l.advance()
		// Check if next character is a digit (negative number)
		if !l.atEnd() && unicode.IsDigit(l.peek()) {
			return l.scanNumber("-", startLine, startColumn), nil
		}
		// Otherwise it's subtraction operator
		return &Token{
			Type:   MINUS,
			Value:  "-",
			Line:   startLine,
			Column: startColumn,
		}, nil
	}

	// Check for operators
	if ch == '+' {
		l.advance()
		return &Token{
			Type:   PLUS,
			Value:  "+",
			Line:   startLine,
			Column: startColumn,
		}, nil
	}

	if ch == '*' {
		l.advance()
		return &Token{
			Type:   MULTIPLY,
			Value:  "*",
			Line:   startLine,
			Column: startColumn,
		}, nil
	}

	if ch == '/' {
		l.advance()
		return &Token{
			Type:   DIVIDE,
			Value:  "/",
			Line:   startLine,
			Column: startColumn,
		}, nil
	}

	// Unknown character
	return nil, &LexerError{
		Message: "Unexpected character '" + string(ch) + "'",
		Line:    startLine,
		Column:  startColumn,
	}
}

// scanNumber scans a numeric literal
func (l *Lexer) scanNumber(prefix string, startLine, startColumn int) *Token {
	value := prefix

	// Scan integer part
	for !l.atEnd() && unicode.IsDigit(l.peek()) {
		value += string(l.advance())
	}

	// Scan decimal part if present
	if !l.atEnd() && l.peek() == '.' {
		value += string(l.advance())
		for !l.atEnd() && unicode.IsDigit(l.peek()) {
			value += string(l.advance())
		}
	}

	return &Token{
		Type:   NUMBER,
		Value:  value,
		Line:   startLine,
		Column: startColumn,
	}
}

// skipWhitespace skips whitespace characters
func (l *Lexer) skipWhitespace() {
	for !l.atEnd() {
		ch := l.peek()
		if ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' {
			l.advance()
		} else {
			break
		}
	}
}

// atEnd checks if we're at the end of input
func (l *Lexer) atEnd() bool {
	return l.pos >= len(l.text)
}

// peek returns the current character without consuming it
func (l *Lexer) peek() rune {
	if l.atEnd() {
		return 0
	}
	return rune(l.text[l.pos])
}

// advance consumes and returns the current character
func (l *Lexer) advance() rune {
	if l.atEnd() {
		return 0
	}
	ch := rune(l.text[l.pos])
	l.pos++
	if ch == '\n' {
		l.line++
		l.column = 1
	} else {
		l.column++
	}
	return ch
}

File: projects/rpn2tex/migrations/java-feature-by-feature-3/src/main/java/com/rpn2tex/Lexer.java
package com.rpn2tex;

import java.util.ArrayList;
import java.util.List;

/**
 * Lexical analyzer for RPN expressions.
 * <p>
 * The lexer scans the input text character-by-character and produces
 * a sequence of tokens. It handles:
 * </p>
 * <ul>
 *   <li>Numeric literals (integers and decimals)</li>
 *   <li>Whitespace (ignored)</li>
 *   <li>Position tracking for error reporting</li>
 * </ul>
 * <p>
 * Example usage:
 * </p>
 * <pre>
 * Lexer lexer = new Lexer("5 3.14");
 * List&lt;Token&gt; tokens = lexer.tokenize();
 * </pre>
 */
public class Lexer {
    private final String text;
    private int position;
    private int line;
    private int column;

    /**
     * Constructs a new Lexer for the given input text.
     *
     * @param text The input text to tokenize
     */
    public Lexer(String text) {
        this.text = text != null ? text : "";
        this.position = 0;
        this.line = 1;
        this.column = 1;
    }

    /**
     * Tokenizes the entire input text.
     *
     * @return A list of tokens, ending with EOF
     * @throws LexerException if an invalid character is encountered
     */
    public List<Token> tokenize() throws LexerException {
        List<Token> tokens = new ArrayList<>();

        while (!atEnd()) {
            Token token = nextToken();
            tokens.add(token);
            if (token.type() == TokenType.EOF) {
                break;
            }
        }

        // Ensure EOF token is present
        if (tokens.isEmpty() || tokens.get(tokens.size() - 1).type() != TokenType.EOF) {
            tokens.add(new Token(TokenType.EOF, "", line, column));
        }

        return tokens;
    }

    /**
     * Scans and returns the next token from the input.
     *
     * @return The next token
     * @throws LexerException if an invalid character is encountered
     */
    private Token nextToken() throws LexerException {
        skipWhitespace();

        if (atEnd()) {
            return new Token(TokenType.EOF, "", line, column);
        }

        char current = peek();
        int startLine = line;
        int startColumn = column;

        // Check for numbers
        if (Character.isDigit(current)) {
            return scanNumber("", startLine, startColumn);
        }

        // Check for addition operator
        if (current == '+') {
            advance();
            return new Token(TokenType.PLUS, "+", startLine, startColumn);
        }

        // Check for minus sign (could be negative number or subtraction operator)
        if (current == '-') {
            advance();
            if (!atEnd() && Character.isDigit(peek())) {
                // It's a negative number (- followed immediately by digit)
                return scanNumber("-", startLine, startColumn);
            }
            // It's a subtraction operator
            return new Token(TokenType.MINUS, "-", startLine, startColumn);
        }

        // Check for multiplication operator
        if (current == '*') {
            advance();
            return new Token(TokenType.TIMES, "*", startLine, startColumn);
        }

        // Check for division operator
        if (current == '/') {
            advance();
            return new Token(TokenType.DIVIDE, "/", startLine, startColumn);
        }

        // Unknown character
        throw new LexerException("Unexpected character: '" + current + "'", startLine, startColumn);
    }

    /**
     * Scans a numeric literal (integer or decimal).
     *
     * @param prefix      The prefix already consumed (e.g., "-" for negative numbers)
     * @param startLine   The line where the number starts
     * @param startColumn The column where the number starts
     * @return A NUMBER token
     */
    private Token scanNumber(String prefix, int startLine, int startColumn) {
        StringBuilder value = new StringBuilder(prefix);

        // Scan integer part
        while (!atEnd() && Character.isDigit(peek())) {
            value.append(advance());
        }

        // Scan decimal part (if present)
        if (!atEnd() && peek() == '.') {
            value.append(advance()); // consume '.'
            while (!atEnd() && Character.isDigit(peek())) {
                value.append(advance());
            }
        }

        return new Token(TokenType.NUMBER, value.toString(), startLine, startColumn);
    }

    /**
     * Skips whitespace characters.
     */
    private void skipWhitespace() {
        while (!atEnd() && Character.isWhitespace(peek())) {
            advance();
        }
    }

    /**
     * Returns the current character without consuming it.
     *
     * @return The current character
     */
    private char peek() {
        if (atEnd()) {
            return '\0';
        }
        return text.charAt(position);
    }

    /**
     * Consumes and returns the current character, advancing position.
     *
     * @return The consumed character
     */
    private char advance() {
        char current = text.charAt(position);
        position++;

        if (current == '\n') {
            line++;
            column = 1;
        } else {
            column++;
        }

        return current;
    }

    /**
     * Checks if we've reached the end of the input.
     *
     * @return true if at end, false otherwise
     */
    private boolean atEnd() {
        return position >= text.length();
    }
}
